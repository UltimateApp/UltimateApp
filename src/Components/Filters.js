import React, { useState, useEffect } from 'react';
import { StyleSheet, View, Text, TouchableOpacity, TextInput } from 'react-native';
import theme from '../styles/theme.style';

const Button = props => {
  const activeStyle = props.active ? styles.activeButton : {};
  return (
    <TouchableOpacity style={{ ...styles.button, ...activeStyle }} onPress={props.onPress}>
      <Text style={styles.buttonText}>{props.title}</Text>
    </TouchableOpacity>
  );
};

class Filters extends React.Component {
  state = {
    selectedLevel: undefined,
    selectedGoal: undefined,
    numberOfPlayers: undefined,
  };

  onPressedChange(target, value) {
    const newValue = value === this.state[target] ? undefined : value;
    this.setState({ [target]: newValue }, this.applyFilters);
  }

  onInputChange(target, value) {
    this.setState({ [target]: value }, this.applyFilters);
  }

  applyFilters() {
    const { selectedLevel, selectedGoal, numberOfPlayers } = this.state;
    let newData = this.props.route.params.initialData;

    if (selectedLevel) newData = newData.filter(drill => drill.level === selectedLevel);
    if (selectedGoal) newData = newData.filter(drill => drill.goals.includes(selectedGoal));
    if (numberOfPlayers) newData = newData.filter(drill => drill.minimalPlayersNumber <= numberOfPlayers);

    this.props.route.params.onFiltered(newData);
  }

  render() {
    const { selectedLevel, selectedGoal, numberOfPlayers } = this.state;
    return (
      <View style={styles.filters}>
        <Text style={styles.filterTitle}>Level</Text>
        <View style={styles.filter}>
          {['beginner', 'intermediate', 'advanced'].map(level => (
            <Button
              key={level}
              title={level}
              onPress={this.onPressedChange.bind(this, 'selectedLevel', level)}
              active={selectedLevel === level}
            />
          ))}
        </View>
        <Text style={styles.filterTitle}>Goals</Text>
        <View style={styles.filter}>
          {this.props.route.params.initialData
            .map(drill => drill.goals)
            .reduce((x, y) => x.concat(y), [])
            .filter((goal, index, array) => array.indexOf(goal) === index)
            .map(goal => (
              <Button
                key={goal}
                title={goal}
                onPress={this.onPressedChange.bind(this, 'selectedGoal', goal)}
                active={selectedGoal === goal}
              />
            ))}
        </View>
        <Text style={styles.filterTitle}>Number of players</Text>
        <TextInput
          style={styles.input}
          placeholder="How many players do you have?"
          onChangeText={this.onInputChange.bind(this, 'numberOfPlayers')}
          value={numberOfPlayers}
          keyboardType="numeric"
        />
        <Button title="Validate" onPress={() => this.props.navigation.goBack()} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  filters: {
    height: '100%',
    alignItems: 'center',
  },
  filterTitle: {
    marginTop: 10,
    textTransform: 'uppercase',
    fontWeight: 'bold',
  },
  filter: {
    marginBottom: 10,
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'center',
  },
  button: {
    borderColor: theme.BORDER_COLOR_BUTTON,
    backgroundColor: theme.BACKGROUND_COLOR_BUTTON,
    borderWidth: 1,
    margin: 5,
    paddingVertical: 10,
    paddingHorizontal: 20,
  },
  activeButton: {
    borderColor: theme.BORDER_COLOR_BUTTON_ACTIVE,
    backgroundColor: theme.BACKGROUND_COLOR_BUTTON_ACTIVE,
  },
  buttonText: {
    textTransform: 'capitalize',
  },
  input: {
    marginBottom: 30,
    padding: 10,
    paddingHorizontal: 20,
    fontSize: 16,
    color: '#444',
    borderBottomWidth: 1,
    borderColor: theme.BORDER_COLOR_BUTTON,
    backgroundColor: theme.BACKGROUND_COLOR_BUTTON,
  },
});

export default Filters;
