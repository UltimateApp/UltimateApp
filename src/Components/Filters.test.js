import React from 'react';
import renderer from 'react-test-renderer';
import { createDrill } from '../Fixtures/TestFixtures';
import { render, fireEvent } from 'react-native-testing-library';
import { Levels } from '../Fixtures';

import Filters from './Filters';

describe('<Filters />', () => {
  const onFiltered = jest.fn();
  const route = {
    params: {
      initialData: [createDrill()],
      onFiltered,
    },
  };
  const navigation = { setOptions: jest.fn() };

  it('renders correctly', () => {
    const tree = renderer.create(<Filters route={route} navigation={navigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('filtering', () => {
    it('filters drills by level', async () => {
      const { container, getByText } = render(<Filters route={route} navigation={navigation} />);

      await fireEvent.press(getByText(Levels.BEGINNER));

      // TODO: Fix tests
      expect(onFiltered).toBeCalledWith([]);

      // await fireEvent.press(getByText('Validate'));

      // expect(onConfirm).toBeCalled();
    });
  });
});
