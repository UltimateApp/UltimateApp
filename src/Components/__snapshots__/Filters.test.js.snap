// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filters /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "fontWeight": "bold",
        "marginTop": 20,
        "textTransform": "uppercase",
      }
    }
  >
    Level
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "#e5e5e5",
          "borderWidth": 1,
          "margin": 5,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text>
        Beginner
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "#e5e5e5",
          "borderWidth": 1,
          "margin": 5,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text>
        Intermediate
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "#e5e5e5",
          "borderWidth": 1,
          "margin": 5,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text>
        Advanced
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderColor": "#e5e5e5",
        "borderWidth": 1,
        "margin": 5,
        "opacity": 1,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
  >
    <Text>
      Validate
    </Text>
  </View>
</View>
`;
